# Importing

In [42]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [44]:
# Importing Products
df_prods = pd.read_csv(r'D:\Career Foundry\2) Data Immersion\4) Python Fundamentals for Data Analysts\DI 4.2 - Jupyter Folder\Instacart Basket Analysis\02 Data\Original Data\4.3_orders_products\products.csv', index_col = False)

In [46]:
# Importing Orders Wrangled
df_ords = pd.read_csv(r'D:\Career Foundry\2) Data Immersion\4) Python Fundamentals for Data Analysts\DI 4.2 - Jupyter Folder\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [47]:
#Path
path=r'D:\Career Foundry\2) Data Immersion\4) Python Fundamentals for Data Analysts\DI 4.2 - Jupyter Folder\Instacart Basket Analysis'

In [50]:
path

'D:\\Career Foundry\\2) Data Immersion\\4) Python Fundamentals for Data Analysts\\DI 4.2 - Jupyter Folder\\Instacart Basket Analysis'

## Data Consistency Checks

In [52]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

## Mix Type Data

In [54]:
# Create a dataframe
df_test = pd.DataFrame()

In [56]:
# Create a Mixed Type Column
df_test['mix']=['a', 'b', 1, True]

In [58]:
# Test
df_test.head()

mix
0     a
1     b
2     1
3  True

In [60]:
# Advance Python Syntax/Function to Search Mixed-Type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\jmesi\AppData\Local\Temp\ipykernel_12488\2908988658.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [14]:
# I think this function worked but I'm unsure of what the note in red means.

In [62]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

## Missing Data

In [64]:
# Finding Missing values of Product Name
df_nan=df_prods[df_prods['product_name'].isnull() == True]

In [66]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [68]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [70]:
# Number of rows of current vs subset
df_prods.shape

(49693, 5)

In [72]:
# New Dataframe with no missing values
df_prods_clean=df_prods[df_prods['product_name'].isnull() == False]

In [74]:
df_prods_clean.shape

(49677, 5)

## Duplicates

In [76]:
# Findings duplicate rows
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [78]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [80]:
# Checking Rows before deleting duplicates
df_prods_clean.shape

(49677, 5)

In [82]:
# Addressing Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [84]:
# Checking that rows were dropped
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

# DI 4.5 - Data Consistency Check Task

In [86]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

## Looking at the chart above there are several things to point out. Looking at the Count row, there are many duplicate numbers being used 3.421083 is the same from unnamed through hour_of_day. The only odd number is the 3.214874 in the since_last_order column. The mean appears to be solid as well as the standard dev. Minimum  values apppear to be low and Max values don't appear to be that high. This leads me to believe that there could be an issue with the numbers somewhere or the data type. Also the max hour_of_day  is 2.3...wouldn't that be 2:00am? That's an odd hour to have orders.

In [88]:
# Checking for Mixed Data in Orders
df_ords.dtypes

Unnamed: 0                 int64
order_id                   int64
user_id                    int64
eval_set                  object
order_number               int64
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

In [90]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_last_order  
0                  8                    NaN  
1                  7                   15.0  
2                 12                   21.0  
3                  7                   29.0  
4                 15                   28.0

In [92]:
df_ords.tail()

Unnamed: 0  order_id  user_id eval_set  order_number  \
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
3421078                   5                 18                   29.0  
3421079                   4                 10                   30.0  
3421080                   1                 12                   18.0  
3421081                   1                 12                    7.0  
3421082                   6                 14                   30.0

In [94]:
# Settine 'eval_set' to String as it's listed as Object
df_ords['eval_set'].astype(str)

0          prior
1          prior
2          prior
3          prior
4          prior
           ...  
3421078    prior
3421079    prior
3421080    prior
3421081    prior
3421082    train
Name: eval_set, Length: 3421083, dtype: object

## I'm lead to believe that Object is usually an indicator of Mixed-Data-Types

In [96]:
# Missing Values
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

## After running the appropriate code my results indicate that there are 206,209 missing values from day_since_last_ordered. There could be various reasons for these missing values such as a company that doesn't order often or is not longer ordering from this business or perhaps they're out of business and data hasn't been updated.

In [98]:
df_na = df_ords[df_ords['days_since_last_order'].isnull() ==True]

In [100]:
df_na

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

In [102]:
df_ords.median()

TypeError: Cannot convert [['prior' 'prior' 'prior' ... 'prior' 'prior' 'train']] to numeric

In [104]:
df_ords['days_since_last_order'].fillna(3, inplace=True)
df_na['days_since_last_order'].fillna(3, inplace=True)

C:\Users\jmesi\AppData\Local\Temp\ipykernel_12488\99873804.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_last_order'].fillna(3, inplace=True)
C:\Users\jmesi\AppData\Local\Temp\ipykernel_12488\99873804.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For

In [106]:
df_na

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    3.0  
11                        2                 11                    3.0  
26                        1                 14                    3.0  
39                        6                 11                    3.0  
45                        3                 12                    3.0  
...                     ...                ...                    ...  
3420930                   4                 12                    3.0  
3420934                   3                 18                    3.0  
3421002                   6                 19                    3.0  
3421019                   1                 15                    3.0  
3421069                   3                 11                    3.0  

[206209 rows x 8 columns]

In [108]:
df_ords.shape

(3421083, 8)

## I chose to change the NaN value of days_since_last_order to 3. This is the Max above in the df.describe() code. I chose this as I thought putting 1 day (mean) would be to often that someone would order form Instacart. 3 days seemed more of an average length of days.

In [110]:
df_ords_clean = df_ords[df_ords['days_since_last_order'].isnull() == False]

In [112]:
df_dupi = df_ords_clean[df_ords_clean.duplicated()]

In [114]:
df_dupi

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

## Checking Products df

In [116]:
# Products Dataframe

df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

## There are no duplicates in the orders.

# Exporting

In [118]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data','orders_clean.csv'))

In [120]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'prods_clean.csv'))